### ISSUE TITLE
Metadata system is all over the place

### ISSUE THREAD

**Issue by globalcitizen** [appc/spec#382]
The so-called 'metadata' system described is relatively complex and has a lack of formal definition around authentication, ie.
```
  "Since the token is used by the Metadata Service to authenticate the pod's identity,
   it SHOULD have no fewer than 128 bits of entropy (i.e. size of UUID), and SHOULD NOT
   be easily guessable (e.g. the pod UUID should not be used)."
```

In addition, since the metadata system involves networking and is a presumably a single instance for all application containers run on a given ACE (application container executor), it creates a network attack vector for container escape to the host, plus a need for shared network address spaces across containers which to some extent negates the very isolation containers are designed to produce and creates a single point of failure within the ACE.

The metadata system explicitly requires the sharing of meta-information (ACE or infrastructure-level UUID information) within the guest, which is of dubious architectural validity (the host can already identify the guest, for example through layer 3 network addresses or kernel container implementation layer identifiers, thus the guest should not have to identify itself!) and creates scope for entropy within
the resulting container which again undermines the determinism that could be achieved.

By exporting the 'fully-reified' manifest to the container through the metadata system, an ACE leaks information about its reification process (ie. trust model). This is clearly undesirable from a security standpoint as well as representing an additional source of container entropy / non-determinism.

By exporting the 'annotations', similar effects are achieved however the definition of annotations is as-yet unclear so further analysis is deferred.

The design of the metadata system seems to be fairly messy also from the perspective of simplicity/clarity/single-point-of-truth: part of the container metadata is shared through the metadata network service, part through environment variables and part through both!

**Comment by eyakubovich** [comment]
Regarding ACE identifying pods without a token. Yes, it is possible and has been like so till very recently (rkt still does that by IP). However, it places a burden on ACE to prevent containers spoofing their IP, removes possibility of container sharing network namespace with host, etc. See https://github.com/appc/spec/pull/327

**Comment by globalcitizen** [comment]
IP spoofing is trivial to prevent using netfilter. You can also prevent frame spoofing.

To be clear, with goals of extensibility/portability/workload isolation you should probably not be using bridges, rather `veth` interfaces (tun/tap) and layer 3 routing. It's also faster since you don't have the spanning tree hassles.

**Comment by eyakubovich** [comment]
I wish we could always just keep everything at layer 3. But having worked on flannel (https://github.com/coreos/flannel), which is at L3, we have received our share of requests for L2:
- Multicast w/o setting up routing (PIM, etc) within just a single subnet
- Overlapping IPs to support multiple tenants
- Using macvlan devices
- Using internal OVS ports

Also, veths are slow (not sure why but that's been constantly reported). There's at least enough to not make L2 difficult to implement (harder than it already is).

**Comment by globalcitizen** [comment]
I think you can make everything stay at layer 3, with a bit of effort.
 * Multicast seems extremely obscure to me as a use case, but IIRC IPv6 should resolve this anyway, so a valid answer might be demand people upgrade to IPv6.
 * Overlapping IPs? I guess what you mean here is efficient allocation with respect to start-of-subnet and end-of-subnet addresses (/31). Basically you want to use every address in a given allocation. You can *certainly* achieve this with a `veth`-style approach without resorting to layer 2 and bridging.
 * I am yet to see any argument for *macvlan* that makes sense. So far it kind of reads to me like "let's think at layer 2 and create more problems for ourselves than we can easily solve, while shooting ourselves in the foot re: potential portability to any other (non-Linux) kernel". Startup times are poor. I've had containers that start so quick, the STP code hasn't finished initializing yet for the interface and the first chunk of traffic just gets lost. Then there was the semi-recent (last 2 years) kernel bug about higher-than-something MAC addresses being allocated and causing networking to fail entirely. This is the sort of random ridiculous pain you get if you start fiddling with layer 2 instead of keeping hardware at arms length and sticking with layer 3. Just stick with layer 3.
 * For OVS, you could have the ACE implement a layer 3 to OVS port bridge, thereby normalizing traffic behavior across OVS and non-OVS implementations.

**Comment by eyakubovich** [comment]
You're right, there're problems with L2 based solutions and I'll be glad to see implementations use L3. However, this is a spec, we want it to be general and allow for L2. One thing we do know is that nobody can agree on the "right way" to virtualize networking and being flexible here is important. And to that end, helping folks out by not also forcing them to be secure against IP spoofing is a nice thing. It also separates the networking and metadata service better -- metadata service can do its own auth and not depend on the network.

**Comment by globalcitizen** [comment]
If you truly want it to be general:
 * Remove the metadata service entirely as it depends on IP
 * Support non-networked containers
 * Make all network configuration interface-oriented
 * Support multiple interfaces
